test/functional/p2p_private_broadcast.py

```diff
@@ -0,0 +1,414 @@
+    "6n36ljyr55szci5ygidmxqer64qr24f4qmnymnbvgehz7qinxnla.b32.i2p",
+    "72yjs6mvlby3ky6mgpvvlemmwq5pfcznrzd34jkhclgrishqdxva.b32.i2p",
+    "a5qsnv3maw77mlmmzlcglu6twje6ttctd3fhpbfwcbpmewx6fczq.b32.i2p",
+    "aovep2pco7v2k4rheofrgytbgk23eg22dczpsjqgqtxcqqvmxk6a.b32.i2p",
+    "bitcoi656nll5hu6u7ddzrmzysdtwtnzcnrjd4rfdqbeey7dmn5a.b32.i2p",
+    "brifkruhlkgrj65hffybrjrjqcgdgqs2r7siizb5b2232nruik3a.b32.i2p",
+    "c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p",
+    "day3hgxyrtwjslt54sikevbhxxs4qzo7d6vi72ipmscqtq3qmijq.b32.i2p",
+    "du5kydummi23bjfp6bd7owsvrijgt7zhvxmz5h5f5spcioeoetwq.b32.i2p",
+    "e55k6wu46rzp4pg5pk5npgbr3zz45bc3ihtzu2xcye5vwnzdy7pq.b32.i2p",
+    "eciohu5nq7vsvwjjc52epskuk75d24iccgzmhbzrwonw6lx4gdva.b32.i2p",
+    "ejlnngarmhqvune74ko7kk55xtgbz5i5ncs4vmnvjpy3l7y63xaa.b32.i2p",
+    "fhzlp3xroabohnmjonu5iqazwhlbbwh5cpujvw2azcu3srqdceja.b32.i2p",
+    "[fc32:17ea:e415:c3bf:9808:149d:b5a2:c9aa]",
+    "[fcc7:be49:ccd1:dc91:3125:f0da:457d:8ce]",
+    "[fcdc:73ae:b1a9:1bf8:d4c2:811:a4c7:c34e]",
+]
+
+
+class P2PPrivateBroadcast(BitcoinTestFramework):
+    def set_test_params(self):
+        self.disable_autoconnect = False
+        self.num_nodes = 2
+
+    def setup_nodes(self):
+        # Start a SOCKS5 proxy server.
+        socks5_server_config = Socks5Configuration()
+        # self.nodes[0] listens on p2p_port(0),
+        # self.nodes[1] listens on p2p_port(1),
+        # thus we tell the SOCKS5 server to listen on p2p_port(self.num_nodes (which is 2))
+        socks5_server_config.addr = ("127.0.0.1", p2p_port(self.num_nodes))
+        socks5_server_config.unauth = True
+        socks5_server_config.auth = True
+
+        self.socks5_server = Socks5Server(socks5_server_config)
+        self.socks5_server.start()
+
+        # Tor ports are the highest among p2p/rpc/tor, so this should be the first available port.
+        ports_base = tor_port(MAX_NODES) + 1
+
+        self.destinations = []
+
+        self.destinations_lock = threading.Lock()
+
+        def destinations_factory(requested_to_addr, requested_to_port):
+            with self.destinations_lock:
+                i = len(self.destinations)
+                actual_to_addr = ""
+                actual_to_port = 0
+                listener = None
+                if i == NUM_INITIAL_CONNECTIONS:
+                    # Instruct the SOCKS5 server to redirect the first private
+                    # broadcast connection from nodes[0] to nodes[1]
+                    actual_to_addr = "127.0.0.1" # nodes[1] listen address
+                    actual_to_port = tor_port(1) # nodes[1] listen port for Tor
+                    self.log.debug(f"Instructing the SOCKS5 proxy to redirect connection i={i} to "
+                                   f"{format_addr_port(actual_to_addr, actual_to_port)} (nodes[1])")
+                else:
+                    # Create a Python P2P listening node and instruct the SOCKS5 proxy to
```

Comment:
How about putting a generic destination factory (without the `i == NUM_INITIAL_CONNECTIONS` special case)  into the test framework (which this test could call in the else branch).

That way, it could be reused by other tests too -  it seem very useful in general to me to be able to have fake nodes from other networks.
For example I am thinking about a functional test for connection management (#27213), where there is no second node, and ended up duplicating this block.

Reply:
Yes. Maybe some stuff from here can be moved to `Socks5Configuration` and/or `Socks5Server`. Maybe the SOCKS5 server can have a default "destinations factory" function which does what is in the `else` branch here. Then there could be callback which either tells where to redirect the Nth connection or tells to use the default

```
default destinations factory:
    # let i be the number of the connection
    if override callback is set:
        whereto = override(i)
        if whereto is not Null:
            redirect to whereto
            end
    redirect to default
```

this originated at https://github.com/bitcoin/bitcoin/pull/29420

Reply:
Marking this as resolved as work on it continues at https://github.com/bitcoin/bitcoin/pull/33954